pkg/cmd/root/help_topic.go

```diff
@@ -262,30 +276,39 @@ var HelpTopics = map[string]map[string]string{
 	},
 }
 
-func NewHelpTopic(ios *iostreams.IOStreams, topic string) *cobra.Command {
+func NewCmdHelpTopic(ios *iostreams.IOStreams, ht helpTopic) *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     topic,
-		Short:   HelpTopics[topic]["short"],
-		Long:    HelpTopics[topic]["long"],
-		Example: HelpTopics[topic]["example"],
+		Use:     ht.name,
+		Short:   ht.short,
+		Long:    ht.long,
+		Example: ht.example,
 		Hidden:  true,
 		Annotations: map[string]string{
 			"markdown:generate": "true",
-			"markdown:basename": "gh_help_" + topic,
+			"markdown:basename": "gh_help_" + ht.name,
 		},
 	}
 
-	cmd.SetHelpFunc(func(c *cobra.Command, args []string) {
-		helpTopicHelpFunc(ios.Out, c, args)
-	})
 	cmd.SetUsageFunc(func(c *cobra.Command) error {
 		return helpTopicUsageFunc(ios.ErrOut, c)
 	})
 
+	helpFunc := func() func(*cobra.Command, []string) {
```

Comment:
I like immediately invoked function expressions to avoid mutating variables, but sometimes people don't like these at all, so let me know.

Reply:
I am not sure what the immediately invoked function expression buys us? We don't tend to use them. I am open to them but if you could explain their purpose a bit more that would be helpful.

Reply:
Oh sure. I feel 2/10 strongly about this, so it's not a big deal. There are sort of three options here:

### Zero value then mutate

```go
var helpFunc func(*cobra.Command, []string)
if ht.helpFuncFn != nil {
    helpFunc = ht.helpFuncFn(ios)
} else {
    helpFunc = func(c *cobra.Command, _ []string) {
	helpTopicHelpFunc(ios.Out, c)
   }
}
```

When I look at this as a maintainer, I am immediately drawn to the zero value. Does it have meaning? Is it definitely set in all branches?

### Default then set

```go
helpFunc := func(c *cobra.Command, _ []string) {
    helpTopicHelpFunc(ios.Out, c)
}

if ht.helpFuncFn != nil {
    helpFunc = ht.helpFuncFn(ios)
} 
```

This answers my questions about the zero value, and I now know that it's always set to some sensible default. Generally, I also tend to favour setting variables exactly once and never changing them again, since when I'm reading code later, it helps with the mental model. Additionally, I find that when things get a bit more complex, it's nice to encapsulate it all inside a function so that I know all set up for this variable belongs inside this function and I don't need to look anywhere else, hence I just default to IIFE:

```go
helpFunc := func() func(*cobra.Command, []string) {
    if ht.helpFuncFn != nil {
        return ht.helpFuncFn(ios)
    }

    return func(c *cobra.Command, _ []string) {
	helpTopicHelpFunc(ios.Out, c)
    }
}()
```

However, I think mutation is generally the [idiomatic way in Go](https://go.dev/doc/faq#Does_Go_have_a_ternary_form) and since you say we don't use them, I'll just change this and commit to that style to avoid introducing extra pattern complexity.

Reply:
Thanks for taking the time to explain this pattern a bit more. I generally think that you will see the default and set pattern the most in our codebase. 

What are your thoughts on something like this:
```go
	helpFunc := func(c *cobra.Command, _ []string) {
		if ht.helpFuncFn != nil {
			ht.helpFuncFn(ios)
			return
		}
		helpTopicHelpFunc(ios.Out, c)
	}
```
Similar to default and set but all the logic is encapsulated in a function and the variable is set only once.

Reply:
I think that's fine too but I generally like pushing conditionals _up the stack_ in a sense because like mutability, it helps with mental heuristics because once the conditional is evaluated once you never have to think about it again (as opposed to each time the function is called here).

It doesn't really matter in this simple case. Nice suggestion that I hadn't really thought about though!

I'm going to change this to default and set (also, nice that we have a shared language for the pattern now). Cheers.